Function definition mutator UI

AnmAtAnm opened the pull request:
WIP: Compiles, tests, and runs.

Changes:
 * Fix bad code on mutator <arg> parsing.
 * `ProcedureManager.mutateProcedure(..)`, including managing Inputs and making variables.
 * Major refactor of `NameManager`s to preserve user inputed case.
   Included in this is a real map backing, sorting in `NameManager`,
   and index based methods have been shuffled into `BasicVariableFieldView.VariableAdapter`.

Needs:
 * Workspace stats creating variables when a function definition is attached.
 * Validate variable fields do the right thing with new NameManager (test pass, and so does a cursory interaction).
 * Alert user on variable name conflict.
 * Arg name field as drop downs (whoops)

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/google/blockly-android/603)
<!-- Reviewable:end -->

RoboErikG commented on blocklylib-core/src/main/java/com/google/blockly/android/ui/fieldview/BasicFieldVariableView.java:
```diff
@@ -225,32 +229,31 @@ public void onChanged() {
-                    insertionIndex = i;
+        private int getIndexForVarName(String expected) {
+            int i = 0;
+            for (String varName : mVars) {
```
You've been converting a lot of direct access loops to iterator loops. These are slightly slower and cause a bit more garbage collection. It's not significant, but I'd prefer you not convert them unless there's a reason to.

RoboErikG commented on blocklylib-core/src/main/java/com/google/blockly/android/ui/fieldview/BasicFieldVariableView.java:
```diff
@@ -290,16 +293,15 @@ int getVariableAction(int index) {
             clear();
-            for (int i = 0; i < mVariableNameManager.size(); i++) {
-                add(mVariableNameManager.get(i));
+            for (String varName : mVars) {
```
ditto

RoboErikG commented on blocklylib-core/src/main/java/com/google/blockly/model/Workspace.java:
```diff
@@ -232,14 +233,10 @@ public void loadWorkspaceContents(InputStream is) throws BlockLoadingException {
         mController.resetWorkspace();
-        for (int i = 0; i < vars.length; i++) {
-            mController.addVariable(vars[i]);
+        for (String varName : vars) {
```
Here too.

AnmAtAnm commented:
Review status: 0 of 43 files reviewed at latest revision, 7 unresolved discussions.

---

*[blocklylib-core/src/main/java/com/google/blockly/android/control/ProcedureManager.java, line 240 at r1](https://reviewable.io:443/reviews/google/blockly-android/603#-KkYGylDmZg_Snw8xUBi-r1-240:-KkZN8dWucTo9RQd0fLx:b-bfigm) ([raw file](https://github.com/google/blockly-android/blob/2ebca88882ac0a50eadb5da9761a5395ba375bbf/blocklylib-core/src/main/java/com/google/blockly/android/control/ProcedureManager.java#L240)):*
<details><summary><i>Previously, RoboErikG wrote…</i></summary><blockquote>

Docs for parameters
</blockquote></details>

Done, with a refactor that includes optional / @Nullable arguments (otherwise using existing values).

---

*[blocklylib-core/src/main/java/com/google/blockly/android/control/ProcedureManager.java, line 260 at r1](https://reviewable.io:443/reviews/google/blockly-android/603#-KkYGylDmZg_Snw8xUBi-r1-260:-KkZO1IDSRmv20f0CIDJ:b-psdd3h) ([raw file](https://github.com/google/blockly-android/blob/2ebca88882ac0a50eadb5da9761a5395ba375bbf/blocklylib-core/src/main/java/com/google/blockly/android/control/ProcedureManager.java#L260)):*
<details><summary><i>Previously, RoboErikG wrote…</i></summary><blockquote>

Should you check if it's valid before you add it to the manager?
</blockquote></details>

Removed the above name manager addition here, because it was also done within `groupAndFireEvents(..)`, where it belonged, after all vars had already been validated.

---

*[blocklylib-core/src/main/java/com/google/blockly/android/control/ProcedureManager.java, line 319 at r1](https://reviewable.io:443/reviews/google/blockly-android/603#-KkYGylDmZg_Snw8xUBi-r1-319:-KkZUk8wSyq-2tT9YvsO:b-db25uu) ([raw file](https://github.com/google/blockly-android/blob/2ebca88882ac0a50eadb5da9761a5395ba375bbf/blocklylib-core/src/main/java/com/google/blockly/android/control/ProcedureManager.java#L319)):*
<details><summary><i>Previously, RoboErikG wrote…</i></summary><blockquote>

It's bumping it away from another connection, not away from its current position. This means it always ends up the same distance from the other connection. It should be fine to make that optional though and bump from the current position.
</blockquote></details>

Acknowledged. Clarified TODO.

---

*[blocklylib-core/src/main/java/com/google/blockly/android/ui/fieldview/BasicFieldVariableView.java, line 175 at r1](https://reviewable.io:443/reviews/google/blockly-android/603#-KkYGylDmZg_Snw8xUBk-r1-175:-KkZUeY0p_qefgK9_iYF:b-lbebu7) ([raw file](https://github.com/google/blockly-android/blob/2ebca88882ac0a50eadb5da9761a5395ba375bbf/blocklylib-core/src/main/java/com/google/blockly/android/ui/fieldview/BasicFieldVariableView.java#L175)):*
<details><summary><i>Previously, RoboErikG wrote…</i></summary><blockquote>

remove?
</blockquote></details>

Removed, and other in test class.

---

*[blocklylib-core/src/main/java/com/google/blockly/android/ui/fieldview/BasicFieldVariableView.java, line 249 at r1](https://reviewable.io:443/reviews/google/blockly-android/603#-KkYGylDmZg_Snw8xUBk-r1-249:-KkZ8EY1vO-4RziA1ieC:b-bqf6ue) ([raw file](https://github.com/google/blockly-android/blob/2ebca88882ac0a50eadb5da9761a5395ba375bbf/blocklylib-core/src/main/java/com/google/blockly/android/ui/fieldview/BasicFieldVariableView.java#L249)):*
<details><summary><i>Previously, RoboErikG wrote…</i></summary><blockquote>

You've been converting a lot of direct access loops to iterator loops. These are slightly slower and cause a bit more garbage collection. It's not significant, but I'd prefer you not convert them unless there's a reason to.
</blockquote></details>

Understood about that trade offs. I also realize the TreeSet replacing the SimpleArrayMap has similar trade-offs.

In this case, SortedSet does not have random access, so iterator is the only way to go.
The alternative would be to make this return an array backed data structure. If there was a sorted variant of the [ArraySet](https://developer.android.com/reference/android/support/v4/util/ArraySet.html), I'd probably use that.

---

*[blocklylib-core/src/main/java/com/google/blockly/android/ui/fieldview/BasicFieldVariableView.java, line 303 at r1](https://reviewable.io:443/reviews/google/blockly-android/603#-KkYGylDmZg_Snw8xUBk-r1-303:-KkZ8I7iyTLxxx-LAmjM:b-9s9lef) ([raw file](https://github.com/google/blockly-android/blob/2ebca88882ac0a50eadb5da9761a5395ba375bbf/blocklylib-core/src/main/java/com/google/blockly/android/ui/fieldview/BasicFieldVariableView.java#L303)):*
<details><summary><i>Previously, RoboErikG wrote…</i></summary><blockquote>

ditto
</blockquote></details>

Same SortedSet mVar as above.

---

*[blocklylib-core/src/main/java/com/google/blockly/model/Workspace.java, line 238 at r1](https://reviewable.io:443/reviews/google/blockly-android/603#-KkYGylDmZg_Snw8xUBo-r1-238:-KkZ8N8A8Xt5Zv1_jLV4:b-wsdjkq) ([raw file](https://github.com/google/blockly-android/blob/2ebca88882ac0a50eadb5da9761a5395ba375bbf/blocklylib-core/src/main/java/com/google/blockly/model/Workspace.java#L238)):*
<details><summary><i>Previously, RoboErikG wrote…</i></summary><blockquote>

Here too.
</blockquote></details>

Again, a set. No random access.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/google/blockly-android/603)*
<!-- Sent from Reviewable.io -->

AnmAtAnm commented:
Awaiting LG

---

Review status: 0 of 43 files reviewed at latest revision, 7 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/google/blockly-android/603#-:-KkdOxSUvNiJDhY7t8-d:b-v026t2)*
<!-- Sent from Reviewable.io -->

RoboErikG commented:
<img class="emoji" title=":lgtm_strong:" alt=":lgtm_strong:" align="absmiddle" src="https://reviewable.io/lgtm_strong.png" height="20" width="61"/> :

---

Review status: 0 of 43 files reviewed at latest revision, 8 unresolved discussions.

---

*[blocklylib-core/src/main/java/com/google/blockly/android/control/ProcedureManager.java, line 255 at r2](https://reviewable.io:443/reviews/google/blockly-android/603#-Kkl0yZUn2pTPPF7D8VJ:-Kkl0yZUn2pTPPF7D8VK:b-950n4x) ([raw file](https://github.com/google/blockly-android/blob/cd0fe8f4eef8c23bf8ed6978fde73c1e4d793ca7/blocklylib-core/src/main/java/com/google/blockly/android/control/ProcedureManager.java#L255)):*
> ```Java
>      *                                  name, or if argument name is invalid.
>      */
>     public void mutateProcedure(final String originalProcedureName,
> ```

@NonNull for the originalProcedureName?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/google/blockly-android/603#-:-Kkl2mYrEOnGWrt2EkxR:b-yncq55)*
<!-- Sent from Reviewable.io -->
